### DISCUSSION TITLE
Client "Managed Flooding" vs "Router" Function Implementation Questions

### DISCUSSION THREAD

**Discussion by carlbatch** [meshtastic/Discussions#117] [Firmware]
Hello. Our Amateur Radio club has deployed about a dozen nodes for emergency communications between members.  We're in a mountainous area of Western NC.  We're not running on Slot 20 to avoid contention with the general Meshtastic community in the Asheville area.   We have a couple of questions:
1.  In Client mode, what are the details of how it chooses when to rebroadcast a received message, specifically how is the delay window is determined?  (we've heard that the delay window is mostly influenced by the strength of the received message with the shortest delay resulting from receiving a  weak signal)
2.  As a Client is waiting to rebroadcast a received message, will it completely abort the process if it hears the same message a 2nd time?
3. We've heard that for Routers, they rebroadcast right away after checking that the Slot is clear of any other signal.  Is that true, or is there also a small random time added to reduce the probability of a collision if two routers decide to transmit at the same time?
4. Is it true that Routers will always rebroadcast, even if they hear the same message more than once (assuming that the hop count hasn't been exceeded)?

Thanks for any answers!

**Comment by ianmcorvidae** [comment]
Sure!

So first of all, as far as the window calculations, https://github.com/meshtastic/firmware/blob/master/src/mesh/RadioInterface.cpp#L258 is a good spot to start looking if you're code-aware. `getTxDelayMsecWeighted` is what decides the window for packets being rebroadcast.

1. is basically answered by the above code, but: based on the received SNR the firmware sets a value called CWSize, which is a value between, currently, 2 and 7 (CWmin and CWmax). Based on this, it calculates a random interval via the formula `delay = (2 * CWmax) + random(0, pow(2, CWsize));`. This is multiplied by a millisecond "slot time" that's calculated to correspond to a full cycle of doing channel activity detection for the current modem preset. Basically: wait out the time period within which routers rebroadcast (see answer to 3 as well here), and then choose a random value between 0 and `2^CWsize` and wait that many multiples of the slot time before rebroadcasting.
2. If a client hears another rebroadcast of a message and it's in the client's list of recently seen packets and not yet rebroadcast, the rebroadcast will be canceled, yes.
3. There is still a random wait for collision avoidance, and it's still based on received SNR, but it's within a much smaller range, only between 0 and `2*CWsize` (hence the `(2 * CWmax)` in the client formula -- that waits out the whole period in which routers might rebroadcast, to increase the chances that the rebroadcast will be canceled as in 2. if there's a router that hears it first).
4. Routers are exempt from the canceling of rebroadcasts when they see a packet an extra time, but they will still only rebroadcast once (the packet to rebroadcast is only put into the wait time/the transmission queue the first time the packet is seen, but for a router it won't be removed when another node's rebroadcast is seen, like it is with clients).

Hopefully those answers make sense!

**Reply by carlbatch** [reply]
Thanks for the prompt and detailed response!  Our group will have lot's to discuss on our weekly Meshtastic Zoom session.
